utils/pdfimages.1
.\" Copyright 1998-2011 Glyph & Cog, LLC
.TH pdfimages 1 "15 August 2011"
.SH NAME
pdfimages \- Portable Document Format (PDF) image extractor
(version 3.03)
.SH SYNOPSIS
.B pdfimages
[options]
.I PDF-file image-root
.SH DESCRIPTION
.B Pdfimages
saves images from a Portable Document Format (PDF) file as Portable
Pixmap (PPM), Portable Bitmap (PBM), Portable Network Graphics (PNG),
Tagged Image File Format (TIFF), JPEG, JPEG2000, or JBIG2 files.
.PP
Pdfimages reads the PDF file
.IR PDF-file ,
scans one or more pages, and writes one file for each image,
.IR image-root - nnn . xxx ,
where
.I nnn
is the image number and
.I xxx
is the image type (.ppm, .pbm, .png, .tif, .jpg, jp2, jb2e, or jb2g).  If
.I PDF-file
is \'-', it reads the PDF file from stdin.
.PP
The default output format is PBM (for monochrome images) or PPM for
non-monochrome. The \-png or \-tiff options change to default output
to PNG or TIFF respectively. If both \-png and \-tiff are specified,
CMYK images will be written as TIFF and all other images will be
written as PNG. In addition the \-j, \-jp2, and \-jbig2 options will
cause JPEG, JPEG2000, and JBIG2, respectively, images in the PDF file
to be written in their native format.
.SH OPTIONS
.TP
.BI \-f " number"
Specifies the first page to scan.
.TP
.BI \-l " number"
Specifies the last page to scan.
.TP
.B \-png
Change the default output format to PNG.
.TP
.B \-tiff
Change the default output format to TIFF.
.TP
.B \-j
Write images in JPEG format as JPEG files instead of the default format. The JPEG file is identical to the JPEG data stored in the PDF.
.TP
.B \-jp2
Write images in JPEG2000 format as JP2 files instead of the default format. The JP2 file is identical to the JPEG2000 data stored in the PDF.
.TP
.B \-jbig2
Write images in JBIG2 format as JBIG2 files instead of the default format. JBIG2 data in PDF is of the embedded type. The embedded type of JBIG2 has an optional separate file containing global data. The embedded data is written with the extension .jb2e and the global data (if available) will be written to the same image number with the extension .jb2g. The content of both these files is identical to the JBIG2 data in the PDF.
.TP
.B \-ccitt
Write images in CCITT format as CCITT files instead of the default
format. The CCITT file is identical to the CCITT data stored in the
PDF. PDF files contain additional parameters specifying
how to decode the CCITT data. These parameters are translated to
fax2tiff input options and written to a .params file with the same image
number. The parameters are:
.RS
.TP
.B \-1
1D Group 3 encoding
.TP
.B \-2
2D Group 3 encoding
.TP
.B \-4
Group 4 encoding
.TP
.B \-A
Beginning of line is aligned on a byte boundary
.TP
.B \-P
Beginning of line is not aligned on a byte boundary
.TP
.B \-X n
The image width in pixels
.TP
.B \-W
Encoding uses 1 for black and 0 for white
.TP
.B \-B
Encoding uses 0 for black and 1 for white
.TP
.B \-M
Input data fills from most significant bit to least significant bit.
.RE
.TP
.BI \-min-height " number"
Images with smaller height will be ignored
.TP
.BI \-min-width " number"
Images with smaller width will be ignored
.TP
.B \-all
Write JPEG, JPEG2000, JBIG2, and CCITT images in their native format. CMYK files are written as TIFF files. All other images are written as PNG files.
This is equivalent to specifying the options \-png \-tiff \-j \-jp2 \-jbig2 \-ccitt.
.TP
.B \-list
Instead of writing the images, list the images along with various information for each image. Do not specify an
.IR image-root
with this option.
.IP
The following information is listed for each image:
.RS
.TP
.B page
the page number containing the image
.TP
.B num
the image number
.TP
.B type
the image type:
.PP
.RS
image - an opaque image
.RE
.RS
mask - a monochrome mask image
.RE
.RS
smask - a soft-mask image
.RE
.RS
stencil - a monochrome mask image used for painting a color or pattern
.RE
.PP
Note: Transparency in images is represented in PDF using a separate image for the image and the mask/smask.
The mask/smask used as part of a transparent image always immediately follows the image in the image list.
.TP
.B width
image width (in pixels)
.TP
.B height
image height (in pixels)
.PP
Note: the image width/height is the size of the embedded image, not the size the image will be rendered at.
.TP
.B color
image color space:
.PP
.RS
gray - Gray
.RE
.RS
rgb - RGB
.RE
.RS
cmyk - CMYK
.RE
.RS
lab - L*a*b
.RE
.RS
icc - ICC Based
.RE
.RS
index - Indexed Color
.RE
.RS
sep - Separation
.RE
.RS
devn - DeviceN
.RE
.TP
.B comp
number of color components
.TP
.B bpc
bits per component
.TP
.B enc
encoding:
.PP
.RS
image - raster image (may be Flate or LZW compressed but does not use an image encoding)
.RE
.RS
jpeg - Joint Photographic Experts Group
.RE
.RS
jp2 - JPEG2000
.RE
.RS
jbig2 - Joint Bi-Level Image Experts Group
.RE
.RS
ccitt - CCITT Group 3 or Group 4 Fax
.RE
.TP
.B interp
"yes" if the interpolation is to be performed when scaling up the image
.TP
.B object ID
the image dictionary object ID (number and generation)
.TP
.B x\-ppi
The horizontal resolution of the image (in pixels per inch) when rendered on the pdf page.
.TP
.B y\-ppi
The vertical resolution of the image (in pixels per inch) when rendered on the pdf page.
.TP
.B size
The size of the embedded image in the pdf file. The following suffixes are used: 'B' bytes, 'K' kilobytes, 'M' megabytes, and 'G' gigabytes.
.TP
.B ratio
The compression ratio of the embedded image.
.RE
.TP
.BI \-opw " password"
Specify the owner password for the PDF file.  Providing this will
bypass all security restrictions.
.TP
.BI \-upw " password"
Specify the user password for the PDF file.
.TP
.B \-p
Include page numbers in output file names.
.TP
.B \-print\-filenames
Print image filenames to stdout.
.TP
.B \-q
Don't print any messages or errors.
.TP
.B \-v
Print copyright and version information.
.TP
.B \-h
Print usage information.
.RB ( \-help
and
.B \-\-help
are equivalent.)
.SH EXIT CODES
The Xpdf tools use the following exit codes:
.TP
0
No error.
.TP
1
Error opening a PDF file.
.TP
2
Error opening an output file.
.TP
3
Error related to PDF permissions.
.TP
99
Other error.
.SH AUTHOR
The pdfimages software and documentation are copyright 1998-2011 Glyph
& Cog, LLC.
.SH "SEE ALSO"
.BR pdfdetach (1),
.BR pdffonts (1),
.BR pdfinfo (1),
.BR pdftocairo (1),
.BR pdftohtml (1),
.BR pdftoppm (1),
.BR pdftops (1),
.BR pdftotext (1)
.BR pdfseparate (1),
.BR pdfsig (1),
.BR pdfunite (1)


NEWS
Release 26.08.0:
        core:
         * GPG based signature improvements
         * Internal code improvements

        utils:
         * pdftohtml: Fix crash when using dataurls. Issue #1751
         * pdfimages: Add min-height and min-width options

        glib:
         * Stop using G_GNUC_CONST in _get_type funcs

        build system:
         * Slight increase in compilation of utils folder
         * Fix -Wunused-command-line-argument when using clang
         * Use cmake modern way to check for linker support

Release 26.07.0:
        core:
         * Remove deprecated DCT and JPX decoders
         * Fix signature regression in some cases. KDE bug #520496
         * Fix added annotations getting lost in some cases. Issue #1732
         * GPG based signature improvements
         * Internal code improvements
         * Fix crashes in malformed documents

        utils:
         * pdfinfo: sanitize output

        cpp:
         * Add render_hint::ignore_paper_color to allow transparent paper

        glib:
         * construct PopplerPage::mutex (fixes crash on macOS). Issue #1743

        qt5:
         * Fix crash in some signature operations. Issue #1737

        qt6:
         * Fix crash in some signature operations. Issue #1737

Release 26.06.0:
        core:
         * Improve annotations initialization thread-safety
         * CairoFontEngine: Properly honor font face index
         * Internal code improvements
         * Fix crashes in malformed documents

Release 26.05.0:
        core:
         * Improve reconstruction of damaged files. Issue #1693
         * PSOutputDev: Remove "pipe as filename" feature
         * PSOutputDev: Respect pre-existing PageSize policies. Issue #1610
         * Internal code improvements
         * Fix crashes in malformed documents

        glib:
         * Improve PopplerPage thread-safety

        utils:
         * pdftotext: Add -remove-hyphens option
         * pdftotext: Do not abort on empty strings. Issue #1699

        build system:
         * Increase minimum required dependency versions to those of Ubuntu 24.04

Release 26.04.0:
        core:
         * Splash: Improve knockout groups rendering. Issue #1538
         * Improve reconstruction of damaged files
         * Performance improvement in files with peculiar Form objects. Issue #1682
         * Fix memory leak if embedding png fails
         * Internal code improvements

        qt5:
         * Fix inverted continuation rect in performMultipleTextSearch

        qt6:
         * Fix inverted continuation rect in performMultipleTextSearch

Release 26.03.0:
        core:
         * Add compression support for stamp annotation images
         * NSS signature backend: Look for Firefox profiles also in XDG config directory
         * GPG signature backend: Fix marking of qualified keys
         * Simplify the form of some ink annotations
         * Speed improvements for some fixes
         * Internal code improvements

        qt5:
         * Replace deprecated Qt::SystemLocaleDate
         * Fix wrong result bottom coordinate when searching across lines

        qt6:
         * Fix wrong result bottom coordinate when searching across lines

        glib:
         * Mark structure_element_iter_new as nullable

        build system:
         * Remove USE_FLOAT cmake option

Release 26.02.0:
        core:
         * Improvements in signature checking
         * Improve rendering of files using the CalGray color space
         * Internal code improvements
         * Fix crashes in malformed documents

        utils:
         * pdftotext: Fix page level bounds calculation in tsv mode

        build system:
         * unmaintained value for some options has been replaced by UnmaintainedWillBeRemovedInJuly2026

Release 26.01.0:
        core:
         * Improve signature compatibiluty on some files
         * Add support for de-facto standard annotation icons
         * Splash: Implement more blending modes for mono1/8
         * Internal code improvements
         * Fix crashes in malformed documents

        utils:
         * pdfinfo: Print alt text in -struct

        cpp:
         * Remove deprecated time functions

        qt:
         * Add reading order option for getText()

Release 25.12.0:
        core:
         * Be less strict about the Page Annots object being correct. Issue #1641
         * Fix rendering of some annotations. Issue #1642
         * TextOuputDev: change default line ending to \n
         * Splash: Performance improvements
         * Ignore color operators when painting a Type3 font with d1
         * Internal code improvements
         * Fix crashes in malformed documents

        utils:
         * pdfsig: specify search order for NSS certificate database on the man page

        build system:
         * Switch to C++23

Release 25.11.0:
        core:
         * NSS Signatures: Tweak the logic that decides which firefox profile to use
         * NSS Signatures: call PORT_GetError() only if the preceding CERT_PKIXVerifyCert() fails
         * Splash: Performance improvements
         * Fix crashes in malformed documents

        glib:
         * Fix ocsp check for signatures validation
         * Fix warning when running glib-mkenums

Release 25.10.0:
        core:
         * Fix image signature getting lost
         * Don't embed substitutions for base14 fonts
         * Form font improvements
         * Handle signatures padded with random data
         * Add feature to Ink annotation to render with multiply blend mode
         * Internal code improvements
         * Fix crashes in malformed documents

        glib:
         * Fix signature text
         * Add feature to Ink annotation to render with multiply blend mode

        cpp:
         * Added embedded_file::unicodeName function

Release 25.09.1:
        core:
         * Internal code improvements

        build system:
         * Fix generated .pc files when using old gpgme

Release 25.09.0:

utils/pdfimages.cc
//========================================================================
//
// pdfimages.cc
//
// Copyright 1998-2003 Glyph & Cog, LLC
//
// Modified for Debian by Hamish Moffatt, 22 May 2002.
//
//========================================================================

//========================================================================
//
// Modified under the Poppler project - http://poppler.freedesktop.org
//
// All changes made under the Poppler project to this file are licensed
// under GPL version 2 or later
//
// Copyright (C) 2007-2008, 2010, 2018, 2022, 2024, 2025 Albert Astals Cid <aacid@kde.org>
// Copyright (C) 2010 Hib Eris <hib@hiberis.nl>
// Copyright (C) 2010 Jakob Voss <jakob.voss@gbv.de>
// Copyright (C) 2012, 2013, 2017 Adrian Johnson <ajohnson@redneon.com>
// Copyright (C) 2013 Suzuki Toshiya <mpsuzuki@hiroshima-u.ac.jp>
// Copyright (C) 2018 Adam Reichold <adam.reichold@t-online.de>
// Copyright (C) 2019, 2021 Oliver Sander <oliver.sander@tu-dresden.de>
// Copyright (C) 2019 Hartmut Goebel <h.goebel@crazy-compilers.com>
// Copyright (C) 2024 Fernando Herrera <fherrera@onirica.com>
// Copyright (C) 2024 Sebastian J. Bronner <waschtl@sbronner.com>
// Copyright (C) 2026 g10 Code GmbH, Author: Sune Stolborg Vuorela <sune@vuorela.dk>
// Copyright (C) 2026 Michael <ttmigueltt@gmail.com>
//
// To see a description of the changes please see the Changelog file that
// came with your tarball or type make ChangeLog if you are building from git
//
//========================================================================

#include "config.h"
#include <poppler-config.h>
#include <cstdio>
#include "parseargs.h"
#include "goo/GooString.h"
#include "GlobalParams.h"
#include "Catalog.h"
#include "Page.h"
#include "PDFDoc.h"
#include "PDFDocFactory.h"
#include "ImageOutputDev.h"
#include "Error.h"
#include "Win32Console.h"

static int firstPage = 1;
static int lastPage = 0;
static bool listImages = false;
static bool enablePNG = false;
static bool enableTiff = false;
static bool dumpJPEG = false;
static bool dumpJP2 = false;
static bool dumpJBIG2 = false;
static bool dumpCCITT = false;
static bool allFormats = false;
static bool pageNames = false;
static bool printFilenames = false;
static char ownerPassword[33] = "\001";
static char userPassword[33] = "\001";
static bool quiet = false;
static bool printVersion = false;
static bool printHelp = false;
static int minHeight = 0;
static int minWidth = 0;

static const ArgDesc argDesc[] = { { .arg = "-f", .kind = argInt, .val = &firstPage, .size = 0, .usage = "first page to convert" },
                                   { .arg = "-l", .kind = argInt, .val = &lastPage, .size = 0, .usage = "last page to convert" },
#if ENABLE_LIBPNG
                                   { .arg = "-png", .kind = argFlag, .val = &enablePNG, .size = 0, .usage = "change the default output format to PNG" },
#endif
#if ENABLE_LIBTIFF
                                   { .arg = "-tiff", .kind = argFlag, .val = &enableTiff, .size = 0, .usage = "change the default output format to TIFF" },
#endif
                                   { .arg = "-j", .kind = argFlag, .val = &dumpJPEG, .size = 0, .usage = "write JPEG images as JPEG files" },
                                   { .arg = "-jp2", .kind = argFlag, .val = &dumpJP2, .size = 0, .usage = "write JPEG2000 images as JP2 files" },
                                   { .arg = "-jbig2", .kind = argFlag, .val = &dumpJBIG2, .size = 0, .usage = "write JBIG2 images as JBIG2 files" },
                                   { .arg = "-ccitt", .kind = argFlag, .val = &dumpCCITT, .size = 0, .usage = "write CCITT images as CCITT files" },
                                   { .arg = "-all", .kind = argFlag, .val = &allFormats, .size = 0, .usage = "equivalent to -png -tiff -j -jp2 -jbig2 -ccitt" },
                                   { .arg = "-list", .kind = argFlag, .val = &listImages, .size = 0, .usage = "print list of images instead of saving" },
                                   { .arg = "-opw", .kind = argString, .val = ownerPassword, .size = sizeof(ownerPassword), .usage = "owner password (for encrypted files)" },
                                   { .arg = "-upw", .kind = argString, .val = userPassword, .size = sizeof(userPassword), .usage = "user password (for encrypted files)" },
                                   { .arg = "-p", .kind = argFlag, .val = &pageNames, .size = 0, .usage = "include page numbers in output file names" },
                                   { .arg = "-print-filenames", .kind = argFlag, .val = &printFilenames, .size = 0, .usage = "print image filenames to stdout" },
                                   { .arg = "-min-height", .kind = argInt, .val = &minHeight, .size = 0, .usage = "images with smaller height will be ignored" },
                                   { .arg = "-min-width", .kind = argInt, .val = &minWidth, .size = 0, .usage = "images with smaller width will be ignored" },
                                   { .arg = "-q", .kind = argFlag, .val = &quiet, .size = 0, .usage = "don't print any messages or errors" },
                                   { .arg = "-v", .kind = argFlag, .val = &printVersion, .size = 0, .usage = "print copyright and version info" },
                                   { .arg = "-h", .kind = argFlag, .val = &printHelp, .size = 0, .usage = "print usage information" },
                                   { .arg = "-help", .kind = argFlag, .val = &printHelp, .size = 0, .usage = "print usage information" },
                                   { .arg = "--help", .kind = argFlag, .val = &printHelp, .size = 0, .usage = "print usage information" },
                                   { .arg = "-?", .kind = argFlag, .val = &printHelp, .size = 0, .usage = "print usage information" },
                                   {} };

int main(int argc, char *argv[])
{
    char *imgRoot = nullptr;
    std::optional<GooString> ownerPW, userPW;

    Win32Console win32Console(&argc, &argv);

    // parse args
    const bool ok = parseArgs(argDesc, &argc, argv);
    if (!ok || (listImages && argc != 2) || (!listImages && argc != 3) || printVersion || printHelp) {
        fprintf(stderr, "pdfimages version %s\n", PACKAGE_VERSION);
        fprintf(stderr, "%s\n", popplerCopyright);
        fprintf(stderr, "%s\n", xpdfCopyright);
        if (!printVersion) {
            printUsage("pdfimages", "<PDF-file> <image-root>", argDesc);
        }
        if (printVersion || printHelp) {
            return 0;
        }
        return 99;
    }
    auto *fileName = new GooString(argv[1]);
    if (!listImages) {
        imgRoot = argv[2];
    }

    // read config file
    globalParams = std::make_unique<GlobalParams>();
    if (quiet) {
        globalParams->setErrQuiet(quiet);
    }

    // open PDF file
    if (ownerPassword[0] != '\001') {
        ownerPW = GooString(ownerPassword);
    }
    if (userPassword[0] != '\001') {
        userPW = GooString(userPassword);
    }
    if (fileName->compare("-") == 0) {
        delete fileName;
        fileName = new GooString("fd://0");
    }

    std::unique_ptr<PDFDoc> doc = PDFDocFactory().createPDFDoc(*fileName, ownerPW, userPW);
    delete fileName;

    if (!doc->isOk()) {
        return 1;
    }

    // check for copy permission
#ifdef ENFORCE_PERMISSIONS
    if (!doc->okToCopy()) {
        error(errNotAllowed, -1, "Copying of images from this document is not allowed.");
        return 3;
    }
#endif

    // get page range
    if (firstPage < 1) {
        firstPage = 1;
    }
    if (firstPage > doc->getNumPages()) {
        error(errCommandLine, -1, "Wrong page range given: the first page ({0:d}) can not be larger then the number of pages in the document ({1:d}).", firstPage, doc->getNumPages());
        return 99;
    }
    if (lastPage < 1 || lastPage > doc->getNumPages()) {
        lastPage = doc->getNumPages();
    }
    if (lastPage < firstPage) {
        error(errCommandLine, -1, "Wrong page range given: the first page ({0:d}) can not be after the last page ({1:d}).", firstPage, lastPage);
        return 99;
    }

    // write image files
    auto *imgOut = new ImageOutputDev(imgRoot, pageNames, listImages);
    imgOut->setMinHeight(minHeight);
    imgOut->setMinWidth(minWidth);
    if (imgOut->isOk()) {
        if (allFormats) {
            imgOut->enablePNG(true);
            imgOut->enableTiff(true);
            imgOut->enableJpeg(true);
            imgOut->enableJpeg2000(true);
            imgOut->enableJBig2(true);
            imgOut->enableCCITT(true);
        } else {
            imgOut->enablePNG(enablePNG);
            imgOut->enableTiff(enableTiff);
            imgOut->enableJpeg(dumpJPEG);
            imgOut->enableJpeg2000(dumpJP2);
            imgOut->enableJBig2(dumpJBIG2);
            imgOut->enableCCITT(dumpCCITT);
        }
        imgOut->enablePrintFilenames(printFilenames);
        doc->displayPages(imgOut, firstPage, lastPage, 72, 72, 0, true, false, false);
    }
    const int exitCode = imgOut->isOk() ? 0 : imgOut->getErrorCode();
    delete imgOut;
    return exitCode;
}
